File: frontend/src/api/governance.ts
import { api } from './client';

export interface GovernanceCoverage {
  source: string;
  complete: boolean;
  warnings: string[];
}

export interface CostBreakdown {
  key: string;
  requests: number;
  input_tokens: number;
  output_tokens: number;
  cost_cny: number;
  cache_hits: number;
  fallbacks: number;
  premium_escalations: number;
  errors: number;
  average_latency_ms: number;
}

export interface CostDashboard {
  coverage: GovernanceCoverage;
  currency: 'CNY';
  requests: number;
  input_tokens: number;
  output_tokens: number;
  cost_cny: number;
  cache_hits: number;
  fallbacks: number;
  premium_escalations: number;
  errors: number;
  average_latency_ms: number;
  by_workspace: CostBreakdown[];
  by_user: CostBreakdown[];
  by_conversation: CostBreakdown[];
  by_provider: CostBreakdown[];
  by_model: CostBreakdown[];
  by_route: CostBreakdown[];
  entries: Array<{
    id: string;
    workspace_id: string;
    trace_id: string;
    user_id?: string;
    conversation_id?: string;
    provider: string;
    model: string;
    route?: string;
    status: string;
    input_tokens: number;
    output_tokens: number;
    cost_cny: number;
    latency_ms: number;
    created_at: string;
  }>;
}

export interface CostFilters {
  from?: string;
  to?: string;
  user_id?: string;
  conversation_id?: string;
  route?: string;
  provider?: string;
  model?: string;
}

export interface TraceSummary {
  trace_id: string;
  route?: string;
  status: string;
  started_at: string;
  duration_ms: number;
  stage_count: number;
  provider?: string;
  model?: string;
  tools: string[];
  has_sql: boolean;
  has_rag: boolean;
  has_agent: boolean;
  has_file: boolean;
  has_vision: boolean;
  artifact_count: number;
  error_code?: string;
}

export interface TraceDashboard {
  coverage: GovernanceCoverage;
  trace_granularity: 'COMPLETION_RECEIPT_LEVEL' | 'STAGE_LEVEL';
  items: TraceSummary[];
}

export interface TraceDetail {
  coverage: GovernanceCoverage;
  trace: TraceSummary;
  stages: Array<{
    stage: string;
    status: string;
    started_at: string;
    duration_ms: number;
    timing_source: string;
    provider?: string;
    model?: string;
    tool?: string;
    sql?: string;
    error_code?: string;
  }>;
}

export interface ModelProviderGovernance {
  provider: string;
  display_name: string;
  model?: string;
  configured: boolean;
  health: string;
  circuit_state: string;
  requests: number;
  errors: number;
  average_latency_ms: number;
  cost_cny: number;
  fallback_rate: number;
  premium_ratio: number;
}

export interface ModelDashboard {
  coverage: GovernanceCoverage;
  pricing_version: string;
  default_routes: Record<string, string[]>;
  providers: ModelProviderGovernance[];
}

export interface EvaluationGovernanceRun {
  id: string;
  source: 'DATABASE' | 'EVIDENCE';
  suite: string;
  version?: string;
  source_sha?: string;
  status: string;
  pass_rate?: number;
  result_accuracy?: number;
  citation_accuracy?: number;
  runtime_calls?: number;
  errors: string[];
  artifacts: string[];
  evidence_sha256?: string;
  executed_at?: string;
}

export interface EvaluationGovernanceDashboard {
  coverage: GovernanceCoverage;
  runs: EvaluationGovernanceRun[];
}

export const governanceApi = {
  cost: (filters: CostFilters = {}) => {
    const query = new URLSearchParams();
    Object.entries(filters).forEach(([key, value]) => {
      if (value) query.set(key, value);
    });
    return api<CostDashboard>(`/governance/cost${query.size ? `?${query.toString()}` : ''}`);
  },
  traces: () => api<TraceDashboard>('/governance/traces'),
  trace: (traceId: string) => api<TraceDetail>(`/governance/traces/${encodeURIComponent(traceId)}`),
  models: () => api<ModelDashboard>('/governance/models'),
  evaluation: () => api<EvaluationGovernanceDashboard>('/governance/evaluation'),
};

File: backend/app/schemas/governance.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class GovernanceCoverage(BaseModel):
    source: str
    complete: bool
    warnings: list[str] = Field(default_factory=list)


class CostLedgerEntry(BaseModel):
    id: str
    workspace_id: str
    user_id: str | None = None
    trace_id: str
    request_id: str | None = None
    conversation_id: str | None = None
    route: str | None = None
    capability: str | None = None
    provider: str
    model: str
    status: str
    input_tokens: int = 0
    cached_input_tokens: int = 0
    output_tokens: int = 0
    cost_cny: float = 0
    latency_ms: int = 0
    cache_hit: bool = False
    fallback_count: int = 0
    premium_escalation: bool = False
    retry_count: int = 0
    error_code: str | None = None
    circuit_state: str | None = None
    pricing_version: str | None = None
    source: str
    created_at: datetime


class CostBreakdown(BaseModel):
    key: str
    requests: int
    input_tokens: int
    output_tokens: int
    cost_cny: float
    cache_hits: int
    fallbacks: int
    premium_escalations: int
    errors: int
    average_latency_ms: float


class CostDashboardResponse(BaseModel):
    coverage: GovernanceCoverage
    currency: Literal["CNY"] = "CNY"
    requests: int
    input_tokens: int
    output_tokens: int
    cost_cny: float
    cache_hits: int
    fallbacks: int
    premium_escalations: int
    errors: int
    average_latency_ms: float
    by_workspace: list[CostBreakdown]
    by_user: list[CostBreakdown]
    by_conversation: list[CostBreakdown]
    by_provider: list[CostBreakdown]
    by_model: list[CostBreakdown]
    by_route: list[CostBreakdown]
    entries: list[CostLedgerEntry]


class TraceStageRead(BaseModel):
    stage: str
    status: str
    started_at: datetime
    duration_ms: int = 0
    timing_source: str
    provider: str | None = None
    model: str | None = None
    tool: str | None = None
    sql: str | None = None
    error_code: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class TraceSummaryRead(BaseModel):
    trace_id: str
    workspace_id: str
    user_id: str | None = None
    route: str | None = None
    status: str
    started_at: datetime
    duration_ms: int = 0
    stage_count: int = 0
    provider: str | None = None
    model: str | None = None
    tools: list[str] = Field(default_factory=list)
    has_sql: bool = False
    has_rag: bool = False
    has_agent: bool = False
    has_file: bool = False
    has_vision: bool = False
    artifact_count: int = 0
    error_code: str | None = None


class TraceDashboardResponse(BaseModel):
    coverage: GovernanceCoverage
    trace_granularity: Literal["COMPLETION_RECEIPT_LEVEL", "STAGE_LEVEL"]
    items: list[TraceSummaryRead]


class TraceDetailResponse(BaseModel):
    coverage: GovernanceCoverage
    trace: TraceSummaryRead
    stages: list[TraceStageRead]


class ModelProviderGovernanceRead(BaseModel):
    provider: str
    display_name: str
    model: str | None = None
    configured: bool
    health: str
    circuit_state: str
    circuit_failure_threshold: int
    circuit_cooldown_seconds: float
    requests: int
    errors: int
    average_latency_ms: float
    cost_cny: float
    fallback_rate: float
    premium_ratio: float


class ModelDashboardResponse(BaseModel):
    coverage: GovernanceCoverage
    pricing_version: str
    default_routes: dict[str, list[str]]
    providers: list[ModelProviderGovernanceRead]


class EvaluationGovernanceRun(BaseModel):
    id: str
    source: Literal["DATABASE", "EVIDENCE"]
    suite: str
    version: str | None = None
    source_sha: str | None = None
    status: str
    pass_rate: float | None = None
    result_accuracy: float | None = None
    citation_accuracy: float | None = None
    runtime_calls: int | None = None
    errors: list[str] = Field(default_factory=list)
    artifacts: list[str] = Field(default_factory=list)
    evidence_sha256: str | None = None
    executed_at: datetime | None = None


class EvaluationGovernanceDashboardResponse(BaseModel):
    coverage: GovernanceCoverage
    runs: list[EvaluationGovernanceRun]


class ModelInvocationContract(BaseModel):
    """Required append-only ONE_MODEL_GATEWAY ledger contract for the shared migration patch."""

    id: str
    workspace_id: str
    user_id: str
    trace_id: str
    request_id: str
    conversation_id: str | None = None
    route: str
    capability: str
    provider: str
    model: str
    status: Literal["SUCCEEDED", "FAILED", "CANCELLED"]
    input_tokens: int = 0
    cached_input_tokens: int = 0
    output_tokens: int = 0
    cost_cny: float = 0
    latency_ms: int = 0
    cache_hit: bool = False
    fallback_count: int = 0
    retry_count: int = 0
    premium_escalation: bool = False
    error_code: str | None = None
    circuit_state: str
    pricing_version: str | None = None
    created_at: datetime
